static/src/javascripts/projects/common/modules/component.spec.js

```diff
@@ -0,0 +1,426 @@
+                ).toBe(true);
+            }
+
+            component.toggleState('state', 'element');
+
+            if (subElem) {
+                expect(
+                    subElem.classList.contains('component__element--state')
+                ).toBe(false);
+            }
+        });
+
+        test('hasState() should return the proper state (without elementName)', () => {
+            component.setState('state');
+            expect(component.hasState('state')).toBe(true);
+            expect(component.hasState('whatever')).toBe(false);
+        });
+
+        test('hasState() should return the proper state (with elementName)', () => {
+            component.setState('state', 'element');
+            expect(component.hasState('state', 'element')).toBe(true);
+            expect(component.hasState('whatever', 'element')).toBe(false);
+        });
+
+        test('hasState() should always return false without componentClass', () => {
+            component.componentClass = '';
+
+            component.setState('state', 'element');
+            expect(component.hasState('state')).toBe(false);
+            expect(component.hasState('state', 'element')).toBe(false);
+        });
+    });
+
+    describe('attachTo()', () => {
+        let elem;
+
+        beforeEach(() => {
+            if (document.body) {
+                document.body.innerHTML = `
+                    <div class="component">
+                        <div class="component__element"></div>
+                    </div>
+                `;
+            }
+
+            elem = document.querySelector('.component');
+
+            if (!elem) {
+                throw new Error('required element not found');
+            }
+        });
+
+        test('calls ready() callback', () => {
+            const component = createComponent({
+                ready: jest.fn(),
+            });
+
+            if (elem) {
+                component.attachTo(elem);
+                expect(component.ready).toBeCalled();
+            }
+        });
+
+        test('calls prerender() callback', () => {
+            const component = createComponent({
+                prerender: jest.fn(),
+            });
+
+            if (elem) {
+                component.attachTo(elem);
+                expect(component.prerender).toBeCalled();
+            }
+        });
+
+        test('calls checkAttached() callback', () => {
+            const component = createComponent({
+                checkAttached: jest.fn(),
+            });
+
+            if (elem) {
+                component.attachTo(elem);
+                expect(component.checkAttached).toBeCalled();
+            }
+        });
+    });
+
+    describe('destroy()', () => {
+        test('calls detach()', () => {
+            const component = createComponent({
+                detach: jest.fn(),
+            });
+
+            component.destroy();
+
+            expect(component.detach).toBeCalled();
```

Comment:
Since `detach` is not designed to be overridable, do you think a more realistic test would ensure that `bean.off(...)` is called?

Reply:
I'm not sure. In this test i wanted to make sure that `destroy()` properly calls `detach()` and whatever happens in there is part of the `detach` test (which is not written yet). So this tried to prove the flow still works as expected. Does this make sense?

Reply:
The trouble with this approach is that we end up testing the implementation of the method rather than the effect it has; how it works, instead of what it does. We don't really care that a method called `detach` is being called, we only care that the event listener is being removed.

I think in part the problem is that exposing the `detach` method is not very helpful, as evidenced by the fact that it is not called externally anywhere in the codebase. Maybe we should just inline `detach`?